python/dlr/tf_model.py

```diff
@@ -0,0 +1,153 @@
+# coding: utf-8
+import os
+import tensorflow as tf
+
+# A prefix that will be prepended to the names in graph_def
+PREFIX = "import"
+
+
+class TFModelImpl:
+    """
+    TFModelImpl is a wrapper on top of tensorflow which implements DLRModel API
+
+    Parameters
+    ----------
+    frozen_graph_file : str
+        Full path to frozen graph (.pb file)
+    dev_type : str
+        Optional. Device type ('cpu' or 'gpu')
+    dev_id : int
+        Optional. Device ID
+    """
+    def __init__(self, frozen_graph_file, dev_type=None, dev_id=None):
+        if not os.path.exists(frozen_graph_file):
+            raise ValueError("frozen_graph_file %s doesn't exist" % frozen_graph_file)
+        self.frozen_graph_file = frozen_graph_file
+
+        device = None
+        if dev_type is not None:
+            devices = ["cpu", "gpu"]
+            if dev_type not in devices:
+                raise ValueError("Invalid device type {}. Valid devices: {}".format(dev_type, devices))
+            dev_id = 0 if dev_id is None else dev_id
+            device = "/{}:{}".format(dev_type, dev_id)
+
+        self.graph = self._load_frozen_graph(frozen_graph_file, device)
+        self.input_tensor_names, self.output_tensor_names = self._get_input_and_output_names(self.graph)
+        self.input_values = {}
+
+    @staticmethod
+    def _load_frozen_graph(frozen_graph_file, device):
+        print("Loading frozen graph: {}".format(frozen_graph_file))
+        with tf.device(device):
+            with tf.gfile.GFile(frozen_graph_file, 'rb') as f:
+                graph_def = tf.GraphDef()
+                graph_def.ParseFromString(f.read())
+            with tf.Graph().as_default() as graph:
+                tf.import_graph_def(graph_def, name=PREFIX)
+            return graph
+
+    @staticmethod
+    def _get_input_and_output_names(graph):
+        input_tensor_names = []
+        output_tensor_names = set()
+        op_prefix = PREFIX + "/"
+        for op in graph.get_operations():
+            if not op.name.startswith(op_prefix):
+                continue
+            if op.type == 'Placeholder' and op.inputs.__len__() == 0 and op.outputs.__len__() == 1:
```

Comment:
len(op.outputs) >= 1 ?

Reply:
I tried to use placeholder Tensor `input1:0` as an input for two other Tensors. When I listed graph_def operations I got `input1:0` operation and its outputs had only 1 item.
I think Placeholder operation type can only have one output tensor.
```
a = tf.placeholder(tf.float32, shape=[2, 2], name="input1")
print("--------------")
print(type(a))
b = tf.placeholder(tf.float32, shape=[2, 2], name="input2")
ab = tf.matmul(a, b)
mm = tf.matmul(a, ab)

for op in graph.get_operations():
    print(op, op.outputs)

name: "import/input1"
op: "Placeholder"
attr {
  key: "dtype"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "shape"
  value {
    shape {
      dim {
        size: 2
      }
      dim {
        size: 2
      }
    }
  }
}
 [<tf.Tensor 'import/input1:0' shape=(2, 2) dtype=float32>]
```

Reply:
got it, i misunderstood outputs here